lib/src/client/session/services/subscriptions/service.rs

```diff
@@ -0,0 +1,984 @@
+        let mut subscription_state = trace_lock!(self.subscription_state);
+        if set_last_publish {
+            subscription_state.set_last_publish();
+        }
+        subscription_state.next_publish_time()
+    }
+
+    /// Send a publish request, returning `true` if the session should send a new request
+    /// immediately.
+    pub(crate) async fn publish(&self) -> Result<bool, StatusCode> {
+        let acks = {
+            let mut subscription_state = trace_lock!(self.subscription_state);
+            let acks = subscription_state.take_acknowledgements();
+            if acks.len() > 0 {
+                Some(acks)
+            } else {
+                None
+            }
+        };
+
+        if log_enabled!(log::Level::Debug) {
+            let sequence_nrs: Vec<u32> = acks
+                .iter()
+                .flatten()
+                .map(|ack| ack.sequence_number)
+                .collect();
+            debug!(
+                "publish is acknowledging subscription acknowledgements with sequence nrs {:?}",
+                sequence_nrs
+            );
+        }
+
+        let request = PublishRequest {
+            request_header: self.channel.make_request_header(self.publish_timeout),
+            subscription_acknowledgements: acks.clone(),
+        };
+
+        let response = self.channel.send(request, self.publish_timeout).await;
+
+        let err_status = match response {
+            Ok(SupportedMessage::PublishResponse(r)) => {
+                session_debug!(self, "PublishResponse");
+
+                let decoding_options = {
+                    let secure_channel = trace_read_lock!(self.channel.secure_channel);
+                    secure_channel.decoding_options()
+                };
+
+                {
+                    let mut subscription_state = trace_lock!(self.subscription_state);
+                    subscription_state.handle_notification(
+                        r.subscription_id,
+                        r.notification_message,
+                        &decoding_options,
+                    );
+                }
+
+                return Ok(r.more_notifications);
+            }
+            Err(e) => e,
+            Ok(r) => {
+                session_error!(self, "publish failed {:?}", r);
+                process_unexpected_response(r)
+            }
+        };
+
+        if let Some(acks) = acks {
+            let mut subscription_state = trace_lock!(self.subscription_state);
+            subscription_state.re_queue_acknowledgements(acks);
+        }
+
+        Err(err_status)
+    }
+
+    /// This code attempts to take the existing subscriptions created by a previous session and
+    /// either transfer them to this session, or construct them from scratch.
+    pub(crate) async fn transfer_subscriptions_from_old_session(&self) {
+        // TODO: This whole thing should probably be optional, so that users can
+        // customize the process.
```

Comment:
Nice `TODO` 😉

I actually just stumbled upon this one as I noticed that each time we had a connection issue, we ended up with an additional subscription. This was caused by this method recreating the existing subscriptions (one of the servers we connect to does not support transfers) and some logic of our own that tries to transfer and/or recreate subscriptions after a connection issue as well 😅

I actually want to manage this myself so I can also reconcile the monitored items (depending on the moment of the interruption and the time it was disconnected it could be the desired items to monitor have changed).

But looking at this (to determine if I could somehow use this instead), I wonder how you can use this method and be sure everything is actually transferred or recreated? It doesn't return any errors so you would still need to query and reconcile all open subscriptions right?

Yet the challenge with that seems to be that if you have multiple subscriptions (which we sometimes have) there is no way to map the old subscription IDs to the new subscription IDs (in case they are recreated).

So wondering how you are doing that (if your doing something like that at all) as the `TODO` says exactly what I was thinking about as solution to the behavior I noticed 😊

Reply:
Really, my thought with that TODO was that advanced users would probably have to do it themselves. It's a bit of an oversight that there's no way to disable it. Part of the reason why I did it this way is that it would take a fair bit more work to add mechanisms to call `TransferSubscriptions` manually, since there is no way to _get_ the actual monitored items from the server, so there's no way to map the subscription IDs as you say.

For you, the best solution would probably be to just have a way to disable it entirely. Maybe we could create some system that would make this approach more reliable.

Reply:
Yeah absolutely... We already have code in place to manage it ourselves so indeed I (for now) just ended up commenting out the line calling this method. Will make a PR to make it optional later this week when I'm done testing.

Yet thinking about making another PR is making me a little bit worried, the number of custom patches in my local fork are becoming a bit challenging to manage in relation to the upstream pending patches 😅

Reply:
Yeah, fully agree... I really respect and appreciate all the time and effort `locka99` is putting/has put into this, so don't want to keep stalking him on issues and PRs 😊 

Yet we are using this crate in some of our production code (well, not the actual crate we use a fork and point Cargo to that repo) so it is important for us to keep this project maintained (either here, which I would prefer, or elsewhere).